### PR TITLE
Bibdk2021 1519 loanerstatus from all libraries

### PR DESCRIPTION
**Pull request by jonaspraem**
Missing tests

### CODE REVIEW

File: src/schema/user.js [hunk 1]
```diff
@@ -60,21 +74,91 @@ function isEmail(email) {
   return /\S+@\S+\.\S+/.test(email);
 }
 
+/**
+ * Filter duplicates, since we get different userIdTypes (LOCAL, CPR)
+ * If both LOCAL and CPR exists, prioritize LOCAL to minimize CPR's sent around services
+ * @param userAccounts: [{ agencyId: String, userId: String, userIdType: String }]
+ * @returns [{ agencyId: String, userId: String, userIdType: String }]
+ */
+const filterDuplicateAgencies = (userAccounts) => {
```
Comment (kiscdbc) [inline src/schema/user.js:83]: jeg laver noget lignende i src/utils/utils.js getUserId, tror jeg 

det vil være fedt at bruge den samme funktion over hele projektet.

File: src/schema/user.js [hunk 2]
```diff
@@ -60,21 +74,91 @@ function isEmail(email) {
...
-      const res = await context.datasources.getLoader("user").load({
+      const userinfo = await context.datasources.getLoader("userinfo").load({
         accessToken: context.accessToken,
       });
+      const homeAccount = getHomeAgencyAccount(userinfo);
+      const res = await context.datasources.getLoader("user").load({
+        userAccount: homeAccount,
+      });
+
       return res.address;
```
Comment (kiscdbc) [inline src/schema/user.js:162]: hvad sker der hvis load fejler? burde der være fejlhåndtering?

**Review by kiscdbc — CHANGES_REQUESTED** [review]

File: src/datasources/debt.datasource.js [hunk 3]
```diff
@@ -1,20 +1,88 @@
...
+    },
+    body: soap,
+  });
+
+  return reduceBody(res?.body, agencyId);
+};
+
+/**
+ * Fetch user debts
+ * @param userAccounts: [{ agencyId: String, userId: String, userIdType: String }]
```
Comment (akri-dbcdk) [inline src/datasources/debt.datasource.js:69]: `userAccounts` er måske ikke det helt rigtige ord, i hvert fald føler jeg ikke at den sidder helt i skabet :thinking: 

Det er det sæt vi får ud af adgangsplatformen, ikke sandt? Kan vi beskrive det lidt mere, måske?

File: src/datasources/debt.datasource.js [hunk 4]
```diff
@@ -1,20 +1,88 @@
...
+        // No loans found, stop here
+        return;
+      }
+      // Add to total list
+      collectedDebts.push(debts);
+    })
+  );
+
+  // Flatten the array
+  return collectedDebts.flat();
```
Comment (akri-dbcdk) [inline src/datasources/debt.datasource.js:87]: Man kan ikke gøre noget i den her stil for at undgå at opdatere variabler løbende?
```
Promise.all(
    userAccounts.map(async (account) => await callService(account, context))
  ).filter((debt) => !!debt)
```

File: src/datasources/loans.datasource.js [hunk 5]
```diff
@@ -1,20 +1,91 @@
...
+        // No loans found, stop here
+        return;
+      }
+      // Add to total list
+      collectedLoans.push(loans);
+    })
+  );
+
+  // Flatten the array
+  return collectedLoans.flat();
```
Comment (akri-dbcdk) [inline src/datasources/loans.datasource.js:90]: Igen. Det er nemmere at følge med, når der ikke ændres/opdateres for meget på variabler løbende så meget

File: src/datasources/loans.datasource.js [hunk 6]
```diff
@@ -1,20 +1,91 @@
...
+    },
+    body: soap,
+  });
+
+  return reduceBody(res?.body, agencyId);
+};
+
+/**
+ * Fetch user loans
+ * @param userAccounts: [{ agencyId: String, userId: String, userIdType: String }]
```
Comment (akri-dbcdk) [inline src/datasources/loans.datasource.js:72]: Igen. UserAccount er ikke det bedste navn måske?

File: src/datasources/orders.datasource.js [hunk 7]
```diff
@@ -1,21 +1,97 @@
...
+
+  return reduceBody(res?.body, agencyId);
+};
+
+/**
+ * Fetch user orders
+ * @param userAccounts: [{ agencyId: String, userId: String, userIdType: String }]
+ */
+export async function load({ userAccounts }, context) {
+  const collectedOrders = [];
```
Comment (akri-dbcdk) [inline src/datasources/orders.datasource.js:81]: Igen, igen

File: src/schema/user.js [hunk 8]
```diff
@@ -11,20 +11,28 @@ import { resolveManifestation } from "../utils/utils";
...
   postalCode: String
   municipalityAgencyId: String
   mail: String
   culrMail: String
+  country: String
   agency(language: LanguageCode): BranchResult!
-  orders: [Order!]!
-  loans: [Loan!]!
-  debt: [Debt!]!
+  orders: [Order!]! @complexity(value: 5)
```
Comment (akri-dbcdk) [inline src/schema/user.js:26]: Jeg tror ikke du behøver at komme complexity på vores interne endpoints, men hvis din story siger noget andet, så bare kør med det

File: src/schema/user.js [hunk 9]
```diff
@@ -84,38 +168,76 @@ export const resolvers = {
       return userinfo?.attributes?.municipalityAgencyId;
     },
     async debt(parent, args, context, info) {
```
Comment (akri-dbcdk) [inline src/schema/user.js:114]: Hvorfor er det egentlig ikke debts, men bare debt? Jeg ved godt at gæld er utælleligt, men har vi ikke kaldt det "debts" flere steder når vi kigger på alle "gælderne" som en bruger har?

File: src/schema/user.js [hunk 10]
```diff
@@ -84,38 +112,76 @@ export const resolvers = {
...
+      const userinfo = await context.datasources.getLoader("userinfo").load({
+        accessToken: context.accessToken,
+      });
+      const homeAccount = getHomeAgencyAccount(userinfo);
+      const res = await context.datasources.getLoader("user").load({
+        userAccount: homeAccount,
+      });
+
+      return res.country;
+    },
```
Comment (akri-dbcdk) [inline src/schema/user.js:184]: Egentlig lidt fjollet at vi både har `user.datasource.js` og `userinfo.datasource.js` :thinking:

File: src/utils/utils.js [hunk 11]
```diff
@@ -413,3 +413,77 @@ export function getUserIdFromAgencyAttributes(agencyAttributes) {
...
+
+  /**
+   * Edge case for test users which municipalityAgencyId doesn't match their accounts
+   */
+  if (!accountLocal && !accountCPR) {
+    // Fall back to any local
+    return userinfo?.attributes?.agencies.find(
+      (account) => account.userIdType === "LOCAL"
+    );
+  }
```
Comment (akri-dbcdk) [inline src/utils/utils.js:451]: Hvorfor sker det her? Og hvorfor afhænger det af `municipalityAgencyId`? Hvorfor lader vi ikke bare Adgangsplatformen om at udstikke id'er, og så være ligeglad med `municipalityAgencyId`?

**Review by akri-dbcdk — CHANGES_REQUESTED** [review]

File: src/utils/utils.js (hunk 11 above)
Comment (jonaspraem) [reply src/utils/utils.js:451]: Hvis en bruger har flere brugere tilkoplet, er vi kun intresseret i den fra deres hjemkommune. Ellers vil openuserstatus returnere gamle data som man er oprettet med tidligere. Fx en gammel addresse.

File: src/schema/user.js (hunk 8 above)
Comment (jonaspraem) [reply src/schema/user.js:26]: Det vel udstillet ligesom alt andet. Selv hvis ingen ente med nogensinde at bruge det, gør det vel ikke noget at holde os selv op på en standard. Det er ihvertfald dyrere end 1

File: src/schema/user.js (hunk 9 above)
Comment (jonaspraem) [reply src/schema/user.js:114]: Ved ikke, den blev kaldt debt af Kasper for 2 år siden, har ikke ændret det her. 

Gæld kan returneres som både en liste eller et samlet gældsoverblik. Tror ikke der er taget ultimativ stilling til det endnu

File: src/schema/user.js (hunk 10 above)
Comment (jonaspraem) [reply src/schema/user.js:184]: Ja. Men user er openuserstatus, userinfo er smaug.

File: src/datasources/debt.datasource.js (hunk 4 above)
Comment (jonaspraem) [reply src/datasources/debt.datasource.js:87]: Krævede lidt modifikation

File: src/schema/user.js (hunk 1 above)
Comment (kiscdbc) [reply src/schema/user.js]: Det kraver lidt modifikation, 
getUserIdFromAgencyAttributes så skal returnere hele accountInfo, så bliver den til getAccountInfoFromAgencyAttributes
Kunne man så ikke:
- filtrer agencies, der matcher homeagencyId
- hvis man ikke finder en match, kalder man bare getAccountInfoFromAgencyAttributes med alle userinfo?.attributes?.agencies
- hvis man finder noget, kalder man getAccountInfoFromAgencyAttributes med homeagency matches

**Review by kiscdbc — APPROVED** [review]

**Review by akri-dbcdk — APPROVED** [review]
Godkendt når Kim også godkender

File: src/__tests__/__snapshots__/user.test.js.snap [hunk 12]
```diff
@@ -3,145 +3,86 @@
...
-          },
-        ],
-      },
-      "mail": "test@dbc.dk",
-      "name": "Freja Damgaard",
-      "postalCode": "8660",
-    },
+    "user": null,
   },
+  "errors": Array [
```
Comment (pjohans) [inline src/__tests__/__snapshots__/user.test.js.snap:8]: denher test er gået i ged - mæhh

File: src/__tests__/__snapshots__/user.test.js.snap [hunk 13]
```diff
@@ -3,145 +3,86 @@
...
```
Comment (pjohans) [inline src/__tests__/__snapshots__/user.test.js.snap:22]: mæhh  siger det lille lam

File: src/__tests__/__snapshots__/user.test.js.snap [hunk 14]
```diff
@@ -3,145 +3,86 @@
...
-          "creator": null,
-          "currency": "DKK",
-          "date": 1969-12-31T23:00:00.000Z,
-          "title": "",
-        },
-      ],
-    },
+    "user": null,
   },
+  "errors": Array [
```
Comment (pjohans) [inline src/__tests__/__snapshots__/user.test.js.snap:51]: same - user.debt er i type sat til IKKE at må være null

File: src/__tests__/__snapshots__/user.test.js.snap [hunk 15]
```diff
@@ -3,145 +3,86 @@
...
+}
+`;
+
+exports[`user - get loans 2`] = `
+Object {
+  "data": Object {
+    "user": null,
+  },
+  "errors": Array [
+    [GraphQLError: Cannot return null for non-nullable field User.loans.],
```
Comment (pjohans) [inline src/__tests__/__snapshots__/user.test.js.snap:74]: jeg tror du skal have mocket hele denher test
